File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
         // Step 1: Build the graph and indegree array
       ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
       int[] inDegree = new int[numCourses];
       
       for(int i = 0; i < numCourses; i++){
           adj.add(new ArrayList<>());
       }
       
       // Populate the adjacency list and indegree array
       for(int[] pre : prerequisites){
           int u = pre[1];  // prerequisite task
            int v = pre[0];  // dependent task

            adj.get(u).add(v);  // Edge from u -> v
            inDegree[v]++;          // Increase indegree of v v has incomming edge u --> v
       }
       
        // Step 2: Initialize the queue with nodes having indegree 0
       Queue<Integer> q = new LinkedList<>();
       for(int i = 0; i < numCourses; i++){
           if(inDegree[i] == 0){
               q.add(i);
           }
       }
       
       // Step 3: Process the queue and perform topological sort
       int count = 0;
       int[] ans = new int[numCourses];
       while(!q.isEmpty()){
           int curr = q.poll();
           ans[count] = curr;
           count++;
           
           // Visit all the neighbors of the node
           for(int neighbor : adj.get(curr)){
               inDegree[neighbor]--;
               
               if(inDegree[neighbor] == 0){
                   q.add(neighbor);
               }
           }
       }

       if(count != numCourses){
        return new int[0];
       }
       // Step 4: If all tasks are processed, return true, else false
       return ans;
    }
}

File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        // Step 1: Build the adjacency list and in-degree array
        vector<vector<int>> adj(numCourses);
        vector<int> inDegree(numCourses, 0);

        for (const auto& pre : prerequisites) {
            int src = pre[1]; // prerequisite task
            int dest = pre[0]; // dependent task

            adj[src].push_back(dest); // Edge from u -> v
            inDegree[dest]++;       // Increase indegree of v (incoming edge u -> v)
        }

        // Step 2: Initialize the queue with nodes having in-degree 0
        queue<int> q;
        for (int i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0) {
                q.push(i);
            }
        }

        // Step 3: Perform topological sort using BFS
        vector<int> ans;
        while (!q.empty()) {
            int curr = q.front();
            q.pop();
            ans.push_back(curr);

            // Visit all the neighbors of the current node
            for (int neighbor : adj[curr]) {
                inDegree[neighbor]--;

                if (inDegree[neighbor] == 0) {
                    q.push(neighbor);
                }
            }
        }

        // Step 4: If all courses are processed, return order; else return an empty vector
        return ans.size() == numCourses ? ans : vector<int>();
    }
};